cms/djangoapps/modulestore_migrator/api/read_api.py

```diff
@@ -0,0 +1,257 @@
+                 ModulestoreMigration table. Please paginate thoughtfully if you do that.
+    """
+    migrations = models.ModulestoreMigration.objects.all().select_related(
+        "source",
+        "target",
+        "target_collection",
+        "task_status",
+    )
+    if source_key:
+        migrations = migrations.filter(source__key=source_key)
+    if target_key:
+        migrations = migrations.filter(target__key=str(target_key))
+    if target_collection_slug:
+        migrations = migrations.filter(target_collection__key=target_collection_slug)
+    if task_uuid:
+        migrations = migrations.filter(task_status__uuid=task_uuid)
+    if is_failed is not None:
+        migrations = migrations.filter(is_failed=is_failed)
+    return (
+        _migration(migration)
+        for migration in migrations.order_by("-id")  # primary key is a proxy for newness
+    )
+
+
+def get_migration_blocks(migration_pk: int) -> dict[UsageKey, ModulestoreBlockMigrationResult]:
+    """
+    Get details about the migrations of each individual block within a course/lib migration.
+    """
+    return {
+        block_migration.source.key: _block_migration_result(block_migration)
+        for block_migration in models.ModulestoreBlockMigration.objects.filter(
+            overall_migration_id=migration_pk
+        ).select_related(
+            "source",
+            "target__learning_package",
+            # For building component key
+            "target__component__component_type",
+            # For building container key
+            "target__container__section",
+            "target__container__subsection",
+            "target__container__unit",
+            # For determining title and version
+            "change_log_record__new_version",
+        )
+    }
+
+
+def _migration(m: models.ModulestoreMigration) -> ModulestoreMigration:
+    """
+    Build a migration dataclass from the database row
+    """
+    return ModulestoreMigration(
+        pk=m.id,
+        source_key=m.source.key,
+        target_key=LibraryLocatorV2.from_string(m.target.key),
+        target_title=m.target.title,
+        target_collection_slug=(m.target_collection.key if m.target_collection else None),
+        target_collection_title=(m.target_collection.title if m.target_collection else None),
+        is_failed=m.is_failed,
+        task_uuid=m.task_status.uuid,
+    )
+
+
+def _block_migration_result(m: models.ModulestoreBlockMigration) -> ModulestoreBlockMigrationResult:
+    """
+    Build an instance of the migration result (successs/failure) dataclass from a database row
+    """
+    return (
+        _block_migration_success(
+            source_key=m.source.key,
+            target=m.target,
+            change_log_record=m.change_log_record,
+        )
+        if m.target
+        else ModulestoreBlockMigrationFailure(
+            source_key=m.source.key,
+            unsupported_reason=(m.unsupported_reason or ""),
+        )
+    )
```

Comment:
Nit: I think that multi-line conditional expressions should generally be avoided, since they can be more confusing to newer Python devs. The more conventional style is just as concise in this case:

```suggestion
    if m.target:
        return _block_migration_success(
            source_key=m.source.key,
            target=m.target,
            change_log_record=m.change_log_record,
        )
    return ModulestoreBlockMigrationFailure(
        source_key=m.source.key,
        unsupported_reason=(m.unsupported_reason or ""),
    )
```